Autopilot/PathManager/Src/landingMode/landingModeStages.cpp

```diff
@@ -0,0 +1,425 @@
+    #else
+        sf_data = land_mode->getSensorFusionData();
+    #endif
+    _input.sensorOutput = &sf_data;
+
+    //making sure landing points are only made once
+    if(!land_mode->getMadeLandingPoints()) {
+        //requires data structure that dhruv wants to use
+        _path = LandingTakeoffManager::createSlopeWaypoints(*_input.telemetryData, _input.sensorOutput->altitude);
+
+        //creating waypoints
+        _path_array[0] = _landing_path.initialize_waypoint(_input.sensorOutput->longitude, _input.sensorOutput->latitude, _input.sensorOutput->altitude, PATH_FOLLOW);
+        _path_array[1] = _landing_path.initialize_waypoint(_path.intersectionPoint.longitude, _path.intersectionPoint.latitude, _path.intersectionPoint.altitude, PATH_FOLLOW);
+        _path_array[2] = _landing_path.initialize_waypoint(_path.aimingPoint.longitude, _path.aimingPoint.latitude, _path.aimingPoint.altitude, PATH_FOLLOW);
+        _path_array[3] = _landing_path.initialize_waypoint(_path.stoppingPoint.longitude, _path.stoppingPoint.latitude, _path.stoppingPoint.altitude, PATH_FOLLOW);
+        _current_location = _landing_path.initialize_waypoint(_input.sensorOutput->longitude, _input.sensorOutput->latitude, _input.sensorOutput->altitude, HOLD_WAYPOINT, 20);
+
+        //initializing flight path
+        _waypoint_status = _landing_path.initialize_flight_path(_path_array, 4, _current_location);
+
+        //set made madelandingPoints to true
+        land_mode->setMadeLandingPoints(true);
+    }
+
+    //translate sensor data to waypoint data in struct
+    _waypoint_input.latitude = _input.sensorOutput->latitude;
+    _waypoint_input.longitude = _input.sensorOutput->longitude;
+    _waypoint_input.altitude = _input.sensorOutput->altitude;
+    _waypoint_input.track = _input.sensorOutput->track;
+
+    //follow the landing waypoints
+    _waypoint_status = _landing_path.get_next_directions(_waypoint_input, &_waypoint_output);
+
+    //translate waypoint commands into landing output structure
+    _output = LandingTakeoffManager::translateWaypointCommands(_waypoint_output);
+
+    //maintaining altitude
+    _output.desiredAirspeed = CRUISING_AIRSPEED;
+
+    //calculating the difference in heading to detect if finished turning (2 differences possible)
+    double difference_in_heading_1 = _input.telemetryData->stoppingDirectionHeading - _input.sensorOutput->track;
+    double difference_in_heading_2 = _input.sensorOutput->track - _input.telemetryData->stoppingDirectionHeading;
+
+    //making sure both headings are positive
+    if(difference_in_heading_1 < 0) {difference_in_heading_1 += 360;}
+    if(difference_in_heading_2 < 0) {difference_in_heading_2 += 360;}
+
+    //if the smaller heading is less than 30 degrees, set stage to slope
+    if((fabs(difference_in_heading_1) <= 80) || (fabs(difference_in_heading_2) <= 80)) {
+        land_mode->setCurrentStage(&LandingSlopeStage::getInstance());
+        land_mode->setCurrentStageEnum(LANDING_SLOPE);
+    }
```

Comment:
Coukd you explain this part of the code a little more please?

Reply:
@NixonCTChan

Reply:
it's almost the same landing code that has been used before for the landing stages, there is documentation on both landing and takeoff if you want to read more about it.

Link: https://uwarg-docs.atlassian.net/wiki/spaces/ZP/pages/1607630921/Landing+and+Takeoff